_posts/2019-02-19-supervisor-strategies.md

```diff
@@ -0,0 +1,368 @@
+Counter.Three: "working and my state is 7"
+Counter.Two: "working and my state is 7"
+Counter.One: "working and my state is 1"
+```
+
+So, we can see our first crash.
+The process for `Counter.One` failed with our raised error, and was restarted.
+Because our default strategy in Elixir is `one_for_one`, this is expected.
+In the default configuration, we dont want one child processes failure to effect any others.
+
+If we let it continue to 22 with `Counter.One`, we would see the same behaviour (allow a crash without impacting any siblings, as its one for one).
+
+## Rest For One
+Now let's try it with `rest_for_one`.
+Rest for one as a strategy starts the children in sequence, and if a later child fails the ones before it do too.
+We want to change our line assigning `opts` in `lib/counter/application.ex` to state that.
+
+```elixir
+# ...
+    children = [
+      Counter.One,
+      Counter.Two,
+      Counter.Three
+    ]
+
+    opts = [strategy: :rest_for_one, name: Counter.Supervisor]
+# ...
+```
+
+Now, let's start up again.
+These logs are also truncated to the interesting part
+
+
+```
+Counter.One: "working and my state is 3"
+Counter.Two: "working and my state is 3"
+Counter.Three: "working and my state is 3"
+Counter.One: "working and my state is 4"
+Counter.Two: "working and my state is 4"
+Counter.Three: "working and my state is 4"
+Counter.One: "working and my state is 5"
+Counter.Two: "working and my state is 5"
+Counter.Three: "starting"
+
+18:30:56.925 [error] GenServer #PID<0.134.0> terminating
+** (RuntimeError) I'm Counter.Three and I'm gonna error now
+    (counter) lib/counter/three.ex:33: Counter.Three.work/1
+    (counter) lib/counter/three.ex:21: Counter.Three.handle_info/2
+    (stdlib) gen_server.erl:616: :gen_server.try_dispatch/4
+    (stdlib) gen_server.erl:686: :gen_server.handle_msg/6
+    (stdlib) proc_lib.erl:247: :proc_lib.init_p_do_apply/3
+Last message: :work
+State: 5
+Counter.Three: "working and my state is 0"
+Counter.Three: "started"
+Counter.One: "working and my state is 6"
+Counter.Two: "working and my state is 6"
+Counter.Three: "working and my state is 0"
+Counter.One: "working and my state is 7"
+Counter.Two: "working and my state is 7"
+```
+
+The key takeaway here is _order matters_.
+Because `Counter.Three` doesn't fail until `22` is its state, and `Counter.One` fails with a state of `5`, `Counter.Three` will force a restart of all 3 children since its last, but `Counter.One`'s failures have no effect on its siblings.
+
+## One For All
+Now let's enable it with `one_for_all`.
+In this supervisory model if one child fails, all must be restarted.
+To do this lets change `lib/counter/application.ex` again.
+
+```elixir
+    opts = [strategy: :one_for_all, name: Counter.Supervisor]
+```
+
+If we start our app again with `iex -S mix` we can see the behaviour as soon as `Counter.Three` reaches a state of 5, but it again will confirm that it works the same again when we reach 22.
```

Comment:
repetition of `again`. Sorry I did not spot this the first time I went through the post.